state/cluster/badger/mutator.go

```diff
@@ -235,6 +229,72 @@ func (m *MutableState) Extend(block *cluster.Block) error {
 	return nil
 }
 
+// checkReferenceBlockValidity validates the reference block is valid.
+//   - it must be a known, finalized block on the main consensus chain
+//   - it must be within the cluster's operating epoch
+//
+// Expected error returns:
+//   - state.InvalidExtensionError if the reference block is invalid for use.
+//   - state.UnverifiableExtensionError if the reference block is unknown.
+func (m *MutableState) checkReferenceBlockValidity(payload *cluster.Payload, finalizedConsensusHeight uint64) error {
+
+	// 1 - the reference block must be known
+	refBlock, err := m.headers.ByBlockID(payload.ReferenceBlockID)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return state.NewUnverifiableExtensionError("cluster block references unknown reference block (id=%x)", payload.ReferenceBlockID)
+		}
+		return fmt.Errorf("could not check reference block: %w", err)
+	}
+
+	// 2 - the reference block must be finalized
+	if refBlock.Height > finalizedConsensusHeight {
+		// a reference block which is above the finalized boundary can't be verified yet
+		return state.NewUnverifiableExtensionError("reference block is above finalized boundary (%d>%d)", refBlock.Height, finalizedConsensusHeight)
+	} else {
+		storedBlockIDForHeight, err := m.headers.BlockIDByHeight(refBlock.Height)
+		if err != nil {
+			return irrecoverable.NewExceptionf("could not look up block ID for finalized height: %w", err)
+		}
+		// a reference block with height at or below the finalized boundary must have been finalized
+		if storedBlockIDForHeight != payload.ReferenceBlockID {
+			return state.NewInvalidExtensionErrorf("cluster block references orphaned reference block (id=%x, height=%d), the block finalized at this height is %x",
+				payload.ReferenceBlockID, refBlock.Height, storedBlockIDForHeight)
+		}
+	}
+
+	// TODO ensure the reference block is part of the main chain
+	_ = refBlock
+
+	var epochFirstHeight uint64
+	var epochLastHeight uint64
+	var epochHasEnded bool
+	m.db.View(func(tx *badger.Txn) error {
```

Comment:
TODO: Make transaction scope in `Extend` much smaller, move this into that transaction

Reply:
https://github.com/onflow/flow-go/pull/4148/commits/8fcf326ef986b98764f1af21dd2fdf3d3579cf94